### ISSUE TITLE
errors with ixnode json submission

### ISSUE THREAD

**Issue by agamba** [ixmaps/ixnode#32]
I've identified some very strange behaviour today when running the ixnode from UofT campus.

A few examples:
https://www.ixmaps.ca/explore.php?trid=77520
![screen shot 2015-06-26 at 3 22 06 pm](https://cloud.githubusercontent.com/assets/999407/8385576/2fba0498-1c17-11e5-9f7f-e279ea672be3.png)

And this same destination using Old TRgGen

![screen shot 2015-06-26 at 3 21 45 pm](https://cloud.githubusercontent.com/assets/999407/8385584/436a907a-1c17-11e5-8445-0bfb12fb468d.png)

The reference data for the IXnode contribution at:
https://www.ixmaps.ca/application/controller/show_tr_contribution.php?tr_c_id=921

My guess is that at some point the raw-socket traceroute encounters an unexpected error, but nonetheless continues the trace but the last reported ip is not updated and somehow lost.

For example see this fragment of the contribution. 
It  continues to report the ip: 128.100.200.243 from hop 7 until hop 24 (which is the MAX_HOPS directive)

Key questions:
* Is the js array containing contributions properly reset after completion of termination of a trset?
* What do the latencies (rtt) mean for the 128.100.200.243?
* Is this a problem of data collection or a more serious issues: e.g. the raw-socket keeps probing the same ip until the end of MAX_HOPS?

----------
tr_c_id: "921",
traceroute_id: "77520",
sub_time: "2015-06-26 12:43:31.236888-04",
dest: "www.claro.com.co",
dest_ip: "184.84.243.34",
city: "Toronto",
country: "CA",
submitter: "lodeanto.ixnode",
submitter_ip: "138.51.248.21",
submitter_os: "Darwin",
postal_code: "UofT",
......

{
pass: "3",
hop: "7",
ttl: "7",
ip: "205.211.94.233",
rtt: "2.67"
},
{
pass: "3",
hop: "8",
ttl: "8",
ip: "128.100.200.243",
rtt: "1.76"
},
{
pass: "3",
hop: "9",
ttl: "9",
ip: "128.100.200.243",
rtt: "2.09"
},
{
pass: "3",
hop: "10",
ttl: "10",
ip: "128.100.200.243",
rtt: "3.85"
},
{
pass: "3",
hop: "11",
ttl: "11",
ip: "128.100.200.243",
rtt: "1.59"
},
{
pass: "3",
hop: "12",
ttl: "12",
ip: "128.100.200.243",
rtt: "1.69"
},
{
pass: "3",
hop: "13",
ttl: "13",
ip: "128.100.200.243",
rtt: "1.75"
},
{
pass: "3",
hop: "14",
ttl: "14",
ip: "128.100.200.243",
rtt: "1.86"
},
{
pass: "3",
hop: "15",
ttl: "15",
ip: "128.100.200.243",
rtt: "1.71"
},
{
pass: "3",
hop: "16",
ttl: "16",
ip: "128.100.200.243",
rtt: "1.65"
},
{
pass: "3",
hop: "17",
ttl: "17",
ip: "128.100.200.243",
rtt: "1.56"
},
{
pass: "3",
hop: "18",
ttl: "18",
ip: "128.100.200.243",
rtt: "1.88"
},
{
pass: "3",
hop: "19",
ttl: "19",
ip: "128.100.200.243",
rtt: "1.62"
},
{
pass: "3",
hop: "20",
ttl: "20",
ip: "128.100.200.243",
rtt: "2.58"
},
{
pass: "3",
hop: "21",
ttl: "21",
ip: "128.100.200.243",
rtt: "1.83"
},
{
pass: "3",
hop: "22",
ttl: "22",
ip: "128.100.200.243",
rtt: "1.81"
},
{
pass: "3",
hop: "23",
ttl: "23",
ip: "128.100.200.243",
rtt: "1.61"
},
{
pass: "3",
hop: "24",
ttl: "24",
ip: "128.100.200.243",
rtt: "2.18"
}

**Comment by agamba** [comment]
Note also how this reveals also the differences in hop count for each of the passes. This an old known issue of the approach we have taken: sending one probe at each time and repeat the trace 4 times, rather than sending 4 probes at each ttl count and running the trace only once.

**Comment by vid** [comment]
Is your second comment relevant to this issue?

It's not clear what you are saying is the bug here. I've seen routers reply with the same address multiple times, just now using traceroute:

        # traceroute www.claro.com.co
        traceroute to www.claro.com.co (184.84.243.34), 30 hops max, 60 byte packets
         1  * * *
         2  192.99.34.252 (192.99.34.252)  0.891 ms  0.901 ms  0.774 ms
         3  192.99.34.252 (192.99.34.252)  0.844 ms  0.865 ms  1.122 ms
         4  mtl-2-6k.qc.ca (198.27.73.6)  1.866 ms  1.939 ms 198.27.73.93 (198.27.73.93)  1.670 ms
         5  mtl-2-6k.qc.ca (198.27.73.6)  2.057 ms  2.172 ms mtl-2-6k.qc.ca (198.27.73.4)  2.179 ms
         6  akamai.ip4.torontointernetxchange.net (206.108.34.24)  8.417 ms  8.423 ms  8.383 ms
         7  akamai.ip4.torontointernetxchange.net (206.108.34.24)  8.386 ms  8.468 ms  8.456 ms
         8  a184-84-243-34.deploy.static.akamaitechnologies.com (184.84.243.34)  8.380 ms  8.347 ms *


I don't think this is a bug with IXnode. Looking at the [raw data](https://www.ixmaps.ca/application/controller/show_tr_contribution.php?tr_c_id=921) each hop is different (ttl, rtt) aside from the IP address. The array is created each time so it would not be using stale data like this.

**Comment by agamba** [comment]
I did run a platform tr from the same origin at the time of the error and did see two hops with the same ip, so you are right that this is not "impossible". I changed location now and this is what I get (which seems more "normal")

Let's look into why are we getting the same ip with different latencies. I will have to look into more detail and run more tests but my first intuition is that in pass 3, after hop 3, the latencies remain in a very close range (which can indicate the raw-socket is probing the same ip)

Here a run from my current (different) location

$ traceroute www.claro.com.co
traceroute: Warning: www.claro.com.co has multiple addresses; using 66.185.85.144
traceroute to a728.b.akamai.net (66.185.85.144), 64 hops max, 52 byte packets
 1  phub.net.cable.rogers.com (192.168.0.1)  4.767 ms  2.489 ms  2.956 ms
 2  7.11.160.41 (7.11.160.41)  8.813 ms  11.665 ms  17.054 ms
 3  209.148.244.45 (209.148.244.45)  29.379 ms  15.630 ms  15.855 ms
 4  69.63.249.229 (69.63.249.229)  10.746 ms  17.839 ms  11.820 ms
 5  69.63.249.58 (69.63.249.58)  20.083 ms  14.264 ms  15.716 ms
 6  * 66.185.85.144 (66.185.85.144)  15.840 ms  16.770 ms

**Comment by agamba** [comment]
Note that ip 128.100.200.243 has ASN: AS239 University of Toronto

I'm not sure where the error might be: 1) raw-socket module, 2) collection of raw-socket data. 
What I'm sure about is that the error or "anomaly" is being submitted in the json file from ixnode

**Comment by agamba** [comment]
I agree, it is possible to have consecutive hops with the same ip address. What strikes me is that the number of responses has the same number of MAX_HOP directive. 

I think my second comment is relevant for this issue but if you consider it's not please do indicate why it's not. We can open a new issue for it. I think it's relevant because the error might be produced due the the approach we are using: 4 passes-1 probe. which as a result produces inconsistent data: what a I mean is passes, each potentially with different number of hops. (If you prefer let's move this to another issue)
______
Note also how this reveals also the differences in hop count for each of the passes. This an old known issue of the approach we have taken: sending one probe at each time and repeat the trace 4 times, rather than sending 4 probes at each ttl count and running the trace only once.
______

Is relevant because

**Comment by agamba** [comment]
another instance
contribution data:
https://www.ixmaps.ca/application/controller/show_tr_contribution.php?tr_c_id=886

![screen shot 2015-06-26 at 4 21 02 pm](https://cloud.githubusercontent.com/assets/999407/8386577/8a5f87e4-1c1f-11e5-82c0-d308047234d8.png)

**Comment by agamba** [comment]
This issue has been resolved by adding a new approach on the server side to analyze inconsistent/differences on data sent from IXmapsClient on each of the attempts/passes
